File: leetcode/binary_tree_max_path_sum/BinaryTreeMaxPathSum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode *root) {
        int maxSum = INT_MIN; 
        
        return max(maxPathSumHelper(root, maxSum), maxSum); 
    }
    
    int maxPathSumHelper(TreeNode * root, int &maxSum) {
        if (root == NULL) {
            return 0; 
        }
        
        int leftSum = maxPathSumHelper(root->left, maxSum); 
        int rightSum = maxPathSumHelper(root->right, maxSum); 
        int returnVal = max(root->val + leftSum, root->val + rightSum);
        returnVal = max(returnVal, root->val); 
        
        maxSum = max(maxSum, leftSum + root->val + rightSum); 
        maxSum = max(maxSum, returnVal); 
        
        return returnVal; 
    }
};

File: leetcode4/BinaryTreeMaxPathSum.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    private int max = Integer.MIN_VALUE; 
    
    public int maxPathSum(TreeNode root) {
        helper(root);
        
        return max; 
    }
    
    private int helper(TreeNode node) {
        if (node == null) {
            return 0; 
        }
        
        int leftMax = helper(node.left);
        int rightMax = helper(node.right); 
        int maxx = Math.max(leftMax, rightMax); 
        maxx = Math.max(node.val, maxx + node.val); 
        max = Math.max(max, maxx); 
        max = Math.max(max, leftMax + node.val + rightMax); 
        
        return maxx; 
    }
}
